### PR TITLE
Server cert tidy

### PR DESCRIPTION
**Pull request by snhenson**
##### Description of change

This PR simplifies server certificate signature algorithm and certificate handling. The function tls_choose_sigalg() is extended to choose the appropriate signature algorithm and certificate for all versions of TLS. For TLS 1.1 and below the signature algorithm is fixed by the ciphersuite.

The corresponding TLS code is updated to use the set signature algorithm instead of having special cases for each version/algorithm.

Some additional simplification will be possible when something similar is done for client auth but that's for another PR.

### CODE REVIEW

File: ssl/ssl_locl.h [hunk 1]
```diff
@@ -1947,6 +1949,13 @@ struct openssl_ssl_test_functions {
 
 const char *ssl_protocol_to_string(int version);
 
+/* Returns true if certificate and private key for 'idx' are present */
+static ossl_inline int ssl_has_cert(const SSL *s, int idx)
+{
+    return s->cert->pkeys[idx].x509 != NULL
```
Comment (richsalz) [inline ssl/ssl_locl.h]: should it check for idx out of range?

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -2135,15 +2129,12 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
...
-    if (!(s->s3->tmp.new_cipher->algorithm_auth & (SSL_aNULL | SSL_aSRP))
-        && !(s->s3->tmp.new_cipher->algorithm_mkey & SSL_PSK)) {
-        if ((pkey = ssl_get_sign_pkey(s, s->s3->tmp.new_cipher, &md))
-            == NULL) {
-            al = SSL_AD_DECODE_ERROR;
-            goto f_err;
-        }
-    } else {
-        pkey = NULL;
+    if ((s->s3->tmp.new_cipher->algorithm_auth & (SSL_aNULL | SSL_aSRP))
```
Comment (richsalz) [inline ssl/statem/statem_srvr.c]: add ```!= 0``` to both tests?

File: ssl/t1_lib.c [hunk 3]
```diff
@@ -784,6 +804,34 @@ static const SIGALG_LOOKUP *tls1_lookup_sigalg(uint16_t sigalg)
...
+        return lu;
+    }
+    return &legacy_rsa_sigalg;
+}
+/* Set peer sigalg based key type */
+int tls1_set_peer_legacy_sigalg(SSL *s, const EVP_PKEY *pkey)
+{
+    int idx = ssl_cert_type(NULL, pkey);
+    const SIGALG_LOOKUP *lu = tls1_get_legacy_sigalg(s, idx);
+    if (lu == NULL)
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after decl.

File: ssl/t1_lib.c [hunk 4]
```diff
@@ -2268,8 +2317,10 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
  * Choose an appropriate signature algorithm based on available certificates
  * Set current certificate and digest to match chosen algorithm.
  */
-int tls_choose_sigalg(SSL *s)
+int tls_choose_sigalg(SSL *s, int *al)
 {
+    int idx;
+    const SIGALG_LOOKUP *lu = NULL;
     if (SSL_IS_TLS13(s)) {
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after decls

File: ssl/ssl_locl.h (hunk 1 above)
Comment (snhenson) [inline ssl/ssl_locl.h]: Well how it's used ensures idx is in range but there is no harm in adding a check: the compiler should optimise most of them away anyway.

**Comment by snhenson** [comment]
Update: that's addressed the comments so far.

**Pull request closed without merging**